pyquil/operator_estimation.py

```diff
@@ -0,0 +1,428 @@
+
+    def __repr__(self):
+        return f'ExperimentResult[{self}]'
+
+    def serializable(self):
+        return {
+            'type': 'ExperimentResult',
+            'experiment': self.experiment,
+            'expectation': self.expectation,
+            'stddev': self.stddev,
+        }
+
+
+def measure_observables(qc: QuantumComputer, experiment_suite: ExperimentSuite, n_shots=1000,
+                        progress_callback=None, active_reset=False):
+    """
+    Measure all the observables in an ExperimentSuite.
+
+    :param qc:
+    :param experiment_suite:
+    :param n_shots:
+    """
+    for i, experiments in enumerate(experiment_suite):
+        log.info(f"Collecting bitstrings for the {len(experiments)} experiments: {experiments}")
+
+        total_prog = Program()
+        already_prepped = dict()
+        for exp in experiments:  # todo: find exp with max weight (for in_operator)
+            for idx, op_str in exp.in_operator:
+                if idx in already_prepped:
+                    assert already_prepped[idx] == op_str
+                else:
+                    total_prog += _local_pauli_eig_prep(op_str, idx)
+                    already_prepped[idx] = op_str
+        total_prog += experiment_suite.program
+
+        already_meased = dict()
+        for exp in experiments:  # todo: find exp with max weight (for out_operator)
+            for idx, op_str in exp.out_operator:
+                if idx in already_meased:
+                    assert already_meased[idx] == op_str
+                else:
+                    total_prog += _local_pauli_eig_meas(op_str, idx)
+                    already_meased[idx] = op_str
+        bitstrings = qc.run_and_measure(total_prog, n_shots, active_reset=active_reset)
+        obs_strings = {q: 1 - 2 * bitstrings[q] for q in bitstrings}
+
+        if progress_callback is not None:
+            progress_callback(i, len(experiment_suite))
+
+        for exp in experiments:
+            measured_qubits = []
+            for idx, op_str in exp.out_operator:
+                measured_qubits.append(idx)
+            measured_qubits = np.array(measured_qubits)
+            log.debug(f"Considering {exp}... 'measuring' qubits {measured_qubits}")
+            assert exp.in_operator.coefficient == 1
+            coeff = exp.out_operator.coefficient
+            if isinstance(coeff, complex):
+                if np.isclose(coeff.imag, 0):
+                    coeff = coeff.real
+
+            if len(measured_qubits) == 0:
+                # I->I
+                yield ExperimentResult(
+                    experiment=exp,
+                    expectation=1.0,
+                    stddev=0.0,
+                )
+                continue
+
+            my_obs_strings = np.vstack(obs_strings[q] for q in measured_qubits).T
+            obs_vals = np.prod(my_obs_strings, axis=1)
+            bit_vals = (1 + obs_vals) // 2
+            n_minus_ones, n_plus_ones = np.bincount(bit_vals, minlength=2)
```

Comment:
Unless deliberate, this step is basically two mistakes cancelling themselves out. `np.bincount(..)` will return the desired tuple in the opposite manner than the one reported above for some arbitrary input array, e.g. `np.bincount([1, 0, 1, 1, 1], minlength=2) returns `array([1, 4])`. But since `bit_vals` (incorrectly, or deliberately) maps `1` to `+1` and `0` to `-1` (usually, we'd do the opposite mapping), this line actually does end up giving the correct counts of `-1`s and `+1`s.

Reply:
oh dear

Reply:
I wrote this in another comment, but I was playing around with different ways of doing variance and settled on using the formula for a beta distribution's variance. But beta distribution is defined on `[0,1]` so I scaled things to this range (and then back) without paying heed to the actual mapping of measured bits to eigenvalues. As you note, it's consistent but weird (and some of the variable names are wrong?). I'll fix this up

Reply:
changed to using off-the-shelf `np.mean` and `np.var`